### PR TITLE
Add fast refresh options

### PR DESCRIPTION
**Pull request by mansam**
If the `ems_openstack.refresh.is_admin` configuration option is set, then ManageIQ will attempt to use admin-only api options to significantly reduce the number of repeated API calls necessary to build the Openstack inventory. This dramatically speeds up the refresh in environments with lots of tenants. (from 9ish minutes in an environment with 100+ tenants, to less than 4 minutes)

Due to differences in the way Heat's access control system works, there is a separate flag to enable admin mode and the associated refresh speedups. This requires Heat to be configured in a particular way, but should further speed up the refresh process.
See https://bugs.launchpad.net/heat/+bug/1466694 and https://bugs.launchpad.net/keystone/+bug/968696 for more details on why this is the case.

### CODE REVIEW

File: lib/manageiq/providers/openstack/legacy/openstack_handle/handled_list.rb [hunk 1]
```diff
@@ -15,7 +16,12 @@ def handled_list(collection_type, options = {})
       # @orchestration_service.handled_list(:resources, :stack => stack)
       # By default, it calls :all method on the Fog collection, that has unified interface in all list methods in Fog
       # and always returns detailed list.
-      multi_tenancy_class.new(self, @os_handle, self.class::SERVICE_NAME, collection_type, options,
+      multi_tenancy_class_ = if all_tenants
+                               OpenstackHandle::MultiTenancy::Option
+                             else
+                               multi_tenancy_class
+                             end
```
Comment (tzumainn) [inline lib/manageiq/providers/openstack/legacy/openstack_handle/handled_list.rb]: The extra _ in the variable seems a bit weird to me - maybe call it 'multi_tenancy_class' and rename the function to 'default_multi_tenancy_class' or something?

File: app/models/manageiq/providers/openstack/inventory/collector/cloud_manager.rb [hunk 2]
```diff
@@ -103,33 +115,47 @@ def quotas
...
+      @tenants ||= identity_service.visible_tenants.select do |t|
+        # avoid 401 Unauth errors when checking for accessible tenants
+        # the "services" tenant is a special tenant in openstack reserved
+        # specifically for the various services
+        next if t.name == "services"
+        true
+      end
+    else
+      @tenants ||= manager.openstack_handle.accessible_tenants
+    end
```
Comment (tzumainn) [inline app/models/manageiq/providers/openstack/inventory/collector/cloud_manager.rb]: Is there a clean way to put the weird 'services' filter into one place code-wise, or is that not really a practical option?

**Comment by tzumainn** [comment]
Thanks!  I put in one or two minor comments/questions.  From an overall perspective: is it reasonable to update the 'classic' refresher code in this PR for backport purposes, or do you think that's better limited to a backport-specific PR?

File: lib/manageiq/providers/openstack/legacy/openstack_handle/handled_list.rb (hunk 1 above)
Comment (mansam) [inline lib/manageiq/providers/openstack/legacy/openstack_handle/handled_list.rb]: Good call.

File: app/models/manageiq/providers/openstack/inventory/collector/cloud_manager.rb (hunk 2 above)
Comment (mansam) [inline app/models/manageiq/providers/openstack/inventory/collector/cloud_manager.rb]: I'll see if I can come up with a way to factor out the repetition cleanly.

**Comment by mansam** [comment]
I think it would be reasonable to do it here.

File: app/models/manageiq/providers/openstack/network_manager/refresh_parser.rb [hunk 3]
```diff
@@ -64,23 +64,43 @@ def parent_manager_find_device_connected_to_network_port(ems_ref)
...
+      end
     end
 
     def floating_ips
-      @network_service.handled_list(:floating_ips)
+      if @ems.kind_of?(ManageIQ::Providers::Openstack::CloudManager) && ::Settings.ems.ems_openstack.refresh.is_admin
+        @floating_ips ||= @network_service.handled_list(:floating_ips, {}, true)
+      else
+        @floating_ips ||= @network_service.handled_list(:floating_ips)
+      end
```
Comment (tzumainn) [inline app/models/manageiq/providers/openstack/network_manager/refresh_parser.rb]: Is it possible to put this code pattern into its own private method, where you pass in :floating_ips or :routers or ... ?

**Comment by tzumainn** [comment]
One more minor comment; I'll test this code out now though.  Thanks!

**Comment by tzumainn** [comment]
Tests look good!  Merging once CI passes

**Comment by simaishi** [comment]
@mansam @tzumainn Can this be `fine/yes` and `euwe/yes` as per flags set on the BZ?

**Comment by mansam** [comment]
@simaishi Portions of this PR apply only to the new inventory object refresh system, but the speedups to the old refresher could be backported to Euwe and Fine.

**Comment by simaishi** [comment]
@mansam in that case... can you please create a PR for Fine/Euwe branch for the applicable changes?

**Comment by mansam** [comment]
@simaishi Will do.

**Comment by simaishi** [comment]
Backported to Euwe via https://github.com/ManageIQ/manageiq/pull/15538

**Comment by simaishi** [comment]
Backported to Fine via https://github.com/ManageIQ/manageiq-gems-pending/pull/224 and https://github.com/ManageIQ/manageiq/pull/15539

**Pull request merged**